Allow use of read-only elfshaker data

mixal-iirec opened the pull request:
With this PR we detect whether repository is readonly.
If so, lockfile and writes to HEAD_FILE are disabled.

This allows usage of `list` and `extract` commands on readonly repositories (#109).

peterwaller-arm commented:
Thank you for your contribution. Apologies for the delay in getting to it. I'm almost inclined to merge it as is.

A few thoughts come to mind:

1) Maybe we can make this work transparently for a readonly data directory (for read operations). That is, if we try to do an operation and it fails because the directory is not writable, fall back to a different path (or report an error if fallback is not possible).

2) This needs tests to defend the functionality. For this it should be as simple as adding a function to check.sh:

https://github.com/elfshaker/elfshaker/blob/9b65a536ed25d698311b7298332900c54967aa97/test-scripts/check.sh#L758-L759

3) Changing the public API is not ideal. We're pre-1.0 so we can still do it, per semantic versioning. But we're about to release 1.0. If it seems plausible it may be necessary to have OpenOptions, then maybe we should introduce such a structure so that the API can be evolved without rewriting consumers of the API to change.

mixal-iirec commented:
1) Automatic fallback sounds like a potentical footgun.
A reader could mistakenly mount a filesystem as readonly (and fallback to different path), then miss synchronization with writers (which would not fallback).
While explicitly using the flag pushes potentical need for external synchronization onto the user.

2) Added test commit.

3) Oh sorry, I somehow didn't realize this is a library as well.
It seems plausible there will be other options that would make sense to be applied to the whole repository - for example the forementioned different path for the HEAD.
It would be better to add such open options pre-1.0. Added commit.

peterwaller-arm commented:
Thanks for the speedy turnaround. Looks reasonable.

> A reader could mistakenly mount a filesystem as readonly (and fallback to different path), then miss synchronization with writers (which would not fallback).

Not sure if I really understand the concern, or if the concern is valid with my view. Are you suggesting there are two mounts of the same repo, one readonly, and one read-write?

The main problematic case I can think of where the lock is needed is where someone might do a GC at the same time as someone else is doing a 'store' or an 'extract'. The GC needs to be able to prevent readers and writers from putting the repostitory into a bad state while objects are in the process of being removed.

I don't really see how having a flag changes the dynamics much. I'm not sure if I'm convinced if a user specifies `--readonly-data` or not is not really going to prevent them from managing to hit the corrupt case where they have both a RO/RW views of the same repository available.

What do you think? I'm thinking broadly that it's better for users to have fewer flags- I'd expect the need for this flag to be rare, and for most use cases if a user has a RO repository they're not going to be writing to it by construction (or doing GC sufficiently rarely or not concurrently with reads/writes) such that this is unlikely to be an issue anyway.

One way the flag could potentially change the game is through documentation: to say that it's not allowed to run a GC on a repostiory which is concurrently being read by another process invoked with `--readonly-data`.

Let me know if I need to re-express any ideas above in case something is unclear.

peterwaller-arm commented:
In the subject you said:
> HEAD, both storing and reading

Not sure what storing means against a readonly repo?

mixal-iirec commented:
> I'm thinking broadly that it's better for users to have fewer flags

I would broadly agree, but I am also a fan of people having to explicitly opt-in into potentially dangerous situations.
Though admittedly this might be unlikely situation, and negative consequences are relatively minor - failure to correctly extract, but the repo itself will be fine.

Ok, flag removed, added detection of readonly repo.



> Not sure what storing means against a readonly repo?

I meant writing of HEAD file - which is otherwise updated with each extract.

peterwaller-arm commented on src/repo/repository.rs:
```diff
@@ -182,15 +182,24 @@ impl Repository {
-                lock_file.try_lock_shared()?;
-            } else {
-                return Err(e.into());
+        let readonly = fs::metadata(&data_dir)?.permissions().readonly();
```
Rather than checking permissions, I'm thinking it would be better to check the error code for file creation. This way you avoid needing an additional check and avoid the possibility that the state changes between the two (admittedly an unlikely scenario here). Also, I'm not sure if `permissions().readonly()` would detect if it were a readonly mount for example.

peterwaller-arm commented:
Glad you came around to my thinking :), it looks like this is nearing completion.

> I meant writing of HEAD file - which is otherwise updated with each extract.

One last thing: Please would you mind updating the PR body so that it says what this PR does. I will squash-merge and use that description (editing it further if I think it's necessary).

mixal-iirec commented:
> Rather than checking permissions, I'm thinking it would be better to check the error code for file creation

I did both - though just as redundancy for the probably too unlikely case where only the lockfile is not writable.

Indeed, `permissions().readonly()` seems unfit for this, removed.

Checking the error code for lockfile creation should be enough.
Though unfortunatelly it seems that `io::ErrorKind` is unstable - bizarrely `io::ErrorKind::ReadOnlyFilesystem` is emitted on stable rust, but directly handling it requires nightly feature flag.
At least this works:
```rust
Err(e) if e.kind().to_string() == "read-only filesystem or storage medium" => None,
```

peterwaller-arm commented:
Thank you for your effort, and thanks for the explanation LGTM. I'm going to merge this now. Please try and use the functionality and report bugs if you encounter them.
